src/Logic.hs

```diff
@@ -483,11 +477,65 @@ pushCandidate (pullRequestId, pullRequest) state = do
 -- effects if it does not change the state.
 proceedUntilFixedPoint :: ProjectState -> Action ProjectState
 proceedUntilFixedPoint state = do
-  newState <- proceed state
+  newState <- provideFeedback state >>= proceed
   if newState == state
     then return state
     else proceedUntilFixedPoint newState
 
+-- Describe the status of the pull request.
+describeStatus :: PullRequestId -> PullRequest -> ProjectState -> Text
+describeStatus prId pr state = case Pr.classifyPullRequest pr of
+  PrStatusAwaitingApproval -> "Pull request awaiting approval."
+  PrStatusApproved ->
+    let approver = fromJust $ Pr.approvedBy pr
+    in case Pr.getQueuePosition prId state of
+      0 -> format "Pull request approved by @{}, rebasing now." [approver]
+      1 -> format "Pull request approved by @{}, waiting for rebase at the front of the queue." [approver]
+      n -> format "Pull request approved by @{}, waiting for rebase behind {} pull requests." (approver, n)
+  PrStatusBuildPending ->
+    let Sha sha = fromJust $ getIntegrationSha pr
+    in Text.concat ["Rebased as ", sha, ", waiting for CI …"]
+  PrStatusIntegrated -> "The build succeeded."
+  PrStatusFailedConflict ->
+    let
+      Branch targetBranchName = fromJust $ getIntegrationTargetBranch pr
+      Branch prBranchName = Pr.branch pr
+    in Text.concat
+      [ "Failed to rebase, please rebase manually using\n\n"
+      , "    git rebase --interactive --autosquash origin/"
+      , targetBranchName
+      , " "
+      , prBranchName
+      ]
+  PrStatusFailedBuild -> "The build failed."
+  where
+    getIntegrationSha :: PullRequest -> Maybe Sha
+    getIntegrationSha pullRequest =
+      case Pr.integrationStatus pullRequest of
+        Integrated sha _ -> Just sha
+        _                -> Nothing
+
+    getIntegrationTargetBranch :: PullRequest -> Maybe Branch
+    getIntegrationTargetBranch pullRequest =
+      case Pr.integrationStatus pullRequest of
+        Conflicted targetBranch -> Just targetBranch
```

Comment:
If it doesn't break other usages of `PullRequestStatus`, we might put the conflicted branch (and also the integrated sha above) in `PrStatusFailedConflict`. I think that would eliminate a few of these impossible branches and `fromJust`.

Reply:
> If it doesn't break other usages

Well, it does, but it wouldn't be too hard to fix.

Reply:
Yeah we need to add the `_` in matches of course, but do we do something like putting them in a `HashMap`? (I don’t recall from the top of my head how the grouping in the webinterface is implemented).

Reply:
I’ll leave it up to you then. We do eventually want to merge the integration and approval status, and at that point `PullRequestStatus` would become obsolete, and we will eliminate many more of these matches that should be unreachable. So we could also leave it like this for now, and fix it as part of the bigger refactor.

Reply:
> but do we do something like putting them in a `HashMap`?

```sh
$ rg HashMap src
```

(No output.)

I guess not :smile:.

Reply:
> So we could also leave it like this for now, and fix it as part of the bigger refactor.

Yeah, I think that's better.

Reply:
Ah this was the part I was worried about: https://github.com/channable/hoff/blob/b21da7705f90f8c87f9b05e3e5ba65deb41f459d/src/WebInterface.hs#L149-L177

We do use `==` but we can add a helper for matching.